File: rustbridge-csharp/RustBridge.Core/MinisignVerifier.cs
using System.Security.Cryptography;
using NSec.Cryptography;

namespace RustBridge;

/// <summary>
/// Minisign signature verification using Ed25519.
/// <para>
/// This class parses and verifies minisign signatures.
/// Minisign uses Ed25519 signatures with a specific file format.
/// </para>
/// <para>
/// <b>Minisign Format:</b>
/// <list type="bullet">
/// <item>Public key: Base64-encoded 42 bytes (2-byte algorithm ID "Ed" + 8-byte key ID + 32-byte key)</item>
/// <item>Signature: Base64-encoded 74 bytes (2-byte algorithm ID "ED" + 8-byte key ID + 64-byte signature)</item>
/// </list>
/// </para>
/// </summary>
/// <seealso href="https://jedisct1.github.io/minisign/">Minisign</seealso>
public class MinisignVerifier
{
    private const int Ed25519PublicKeyBytes = 32;
    private const int Ed25519SignatureBytes = 64;
    private const int KeyIdBytes = 8;
    private const int AlgorithmIdBytes = 2;

    // Algorithm ID for Ed25519 public key ("Ed" = 0x45, 0x64)
    private static readonly byte[] Ed25519PubkeyAlgorithmId = { 0x45, 0x64 };

    // Algorithm ID for Ed25519 signature ("ED" = 0x45, 0x44)
    private static readonly byte[] Ed25519SigAlgorithmId = { 0x45, 0x44 };

    private readonly PublicKey _publicKey;
    private readonly byte[] _keyId;

    /// <summary>
    /// Create a verifier from a minisign public key string.
    /// </summary>
    /// <param name="publicKeyBase64">Minisign public key in base64 format (e.g., "RWS...").</param>
    /// <exception cref="ArgumentNullException">If publicKeyBase64 is null.</exception>
    /// <exception cref="ArgumentException">If the key format is invalid.</exception>
    public MinisignVerifier(string publicKeyBase64)
    {
        ArgumentNullException.ThrowIfNull(publicKeyBase64);
        if (string.IsNullOrWhiteSpace(publicKeyBase64))
        {
            throw new ArgumentException("Public key cannot be empty or whitespace", nameof(publicKeyBase64));
        }

        var (publicKeyBytes, keyId) = ParsePublicKey(publicKeyBase64);
        _keyId = keyId;

        // Create NSec public key from raw bytes
        if (!PublicKey.TryImport(SignatureAlgorithm.Ed25519, publicKeyBytes, KeyBlobFormat.RawPublicKey, out var key) || key == null)
        {
            throw new ArgumentException("Failed to import Ed25519 public key");
        }
        _publicKey = key;
    }

    /// <summary>
    /// Parse a minisign public key from base64 format.
    /// Format: 2 bytes algorithm ID + 8 bytes key ID + 32 bytes public key
    /// </summary>
    private static (byte[] publicKey, byte[] keyId) ParsePublicKey(string publicKeyBase64)
    {
        byte[] decoded;
        try
        {
            decoded = Convert.FromBase64String(publicKeyBase64.Trim());
        }
        catch (FormatException ex)
        {
            throw new ArgumentException("Invalid base64 encoding in public key", ex);
        }

        var expectedLength = AlgorithmIdBytes + KeyIdBytes + Ed25519PublicKeyBytes;
        if (decoded.Length != expectedLength)
        {
            throw new ArgumentException(
                $"Invalid public key length: expected {expectedLength}, got {decoded.Length}");
        }

        // Verify algorithm ID
        var algorithmId = decoded.AsSpan(0, AlgorithmIdBytes);
        if (!algorithmId.SequenceEqual(Ed25519PubkeyAlgorithmId))
        {
            throw new ArgumentException(
                $"Invalid algorithm ID: expected Ed25519, got {Convert.ToHexString(algorithmId)}");
        }

        // Extract the 8-byte key ID (right after algorithm ID)
        var keyId = decoded.AsSpan(AlgorithmIdBytes, KeyIdBytes).ToArray();

        // Extract the 32-byte Ed25519 public key (after key ID)
        var publicKey = decoded.AsSpan(AlgorithmIdBytes + KeyIdBytes, Ed25519PublicKeyBytes).ToArray();

        return (publicKey, keyId);
    }

    /// <summary>
    /// Parse a minisign signature from the multi-line format.
    /// <para>
    /// Format:
    /// <code>
    /// untrusted comment: &lt;comment&gt;
    /// &lt;base64-encoded signature&gt;
    /// trusted comment: &lt;comment&gt;
    /// &lt;base64-encoded global signature&gt;
    /// </code>
    /// </para>
    /// We only use the second line (the signature itself).
    /// </summary>
    private static (byte[] keyId, byte[] signature, bool isPrehashed) ParseSignature(string signatureString)
    {
        var lines = signatureString.Split('\n');
        if (lines.Length < 2)
        {
            throw new CryptographicException("Invalid signature format: expected at least 2 lines");
        }

        // The signature is on the second line
        var signatureBase64 = lines[1].Trim();

        byte[] decoded;
        try
        {
            decoded = Convert.FromBase64String(signatureBase64);
        }
        catch (FormatException ex)
        {
            throw new CryptographicException("Invalid base64 encoding in signature", ex);
        }

        var expectedLength = AlgorithmIdBytes + KeyIdBytes + Ed25519SignatureBytes;
        if (decoded.Length != expectedLength)
        {
            throw new CryptographicException(
                $"Invalid signature length: expected {expectedLength}, got {decoded.Length}");
        }

        // Check algorithm ID - "ED" = prehashed, "Ed" = legacy non-prehashed
        var algorithmId = decoded.AsSpan(0, AlgorithmIdBytes);
        bool isPrehashed;
        if (algorithmId.SequenceEqual(Ed25519SigAlgorithmId))
        {
            isPrehashed = true;  // "ED" - prehashed with BLAKE2b
        }
        else if (algorithmId.SequenceEqual(Ed25519PubkeyAlgorithmId))
        {
            isPrehashed = false; // "Ed" - legacy non-prehashed
        }
        else
        {
            throw new CryptographicException(
                $"Invalid algorithm ID in signature: expected Ed25519, got {Convert.ToHexString(algorithmId)}");
        }

        // Extract key ID
        var keyId = decoded.AsSpan(AlgorithmIdBytes, KeyIdBytes).ToArray();

        // Extract signature
        var signature = decoded.AsSpan(AlgorithmIdBytes + KeyIdBytes, Ed25519SignatureBytes).ToArray();

        return (keyId, signature, isPrehashed);
    }

    /// <summary>
    /// Verify a minisign signature against data.
    /// </summary>
    /// <param name="data">The data that was signed.</param>
    /// <param name="signatureString">The minisign signature (multi-line format).</param>
    /// <returns>True if the signature is valid, false otherwise.</returns>
    /// <exception cref="CryptographicException">If signature parsing fails.</exception>
    public bool Verify(byte[] data, string signatureString)
    {
        return Verify((ReadOnlySpan<byte>)data, signatureString);
    }

    /// <summary>
    /// Verify a minisign signature against data.
    /// </summary>
    /// <param name="data">The data that was signed.</param>
    /// <param name="signatureString">The minisign signature (multi-line format).</param>
    /// <returns>True if the signature is valid, false otherwise.</returns>
    /// <exception cref="CryptographicException">If signature parsing fails.</exception>
    public bool Verify(ReadOnlySpan<byte> data, string signatureString)
    {
        var (sigKeyId, signature, isPrehashed) = ParseSignature(signatureString);

        // Verify key ID matches
        if (!sigKeyId.AsSpan().SequenceEqual(_keyId))
        {
            return false;
        }

        // Minisign "ED" signatures are prehashed - compute BLAKE2b-512 hash first
        // This matches SIGALG_PREHASHED in the minisign crate
        if (isPrehashed)
        {
            var hash = NSec.Cryptography.HashAlgorithm.Blake2b_512.Hash(data);
            return SignatureAlgorithm.Ed25519.Verify(_publicKey, hash, signature);
        }

        // Non-prehashed: verify directly without allocation
        return SignatureAlgorithm.Ed25519.Verify(_publicKey, data, signature);
    }
}

File: rustbridge-python/rustbridge/core/minisign_verifier.py
"""Minisign signature verification using Ed25519."""

from __future__ import annotations

import base64
import hashlib

from nacl.signing import VerifyKey
from nacl.exceptions import BadSignatureError


# Minisign format constants
_ED25519_PUBLIC_KEY_BYTES = 32
_ED25519_SIGNATURE_BYTES = 64
_KEY_ID_BYTES = 8
_ALGORITHM_ID_BYTES = 2

# Algorithm ID for Ed25519 public key ("Ed" = 0x45, 0x64)
_ED25519_PUBKEY_ALGORITHM_ID = bytes([0x45, 0x64])

# Algorithm ID for Ed25519 signature ("ED" = 0x45, 0x44)
_ED25519_SIG_ALGORITHM_ID = bytes([0x45, 0x44])


class MinisignVerifier:
    """
    Minisign signature verification using Ed25519.

    This class parses and verifies minisign signatures.
    Minisign uses Ed25519 signatures with a specific file format.

    Minisign Format:
    - Public key: Base64-encoded 42 bytes (2-byte algorithm ID "Ed" + 8-byte key ID + 32-byte key)
    - Signature: Base64-encoded 74 bytes (2-byte algorithm ID "ED" + 8-byte key ID + 64-byte signature)

    Example:
        verifier = MinisignVerifier("RWS...")
        if verifier.verify(data, signature_content):
            print("Signature valid")
    """

    def __init__(self, public_key_base64: str) -> None:
        """
        Create a verifier from a minisign public key string.

        Args:
            public_key_base64: Minisign public key in base64 format (e.g., "RWS...").

        Raises:
            ValueError: If the key format is invalid.
        """
        public_key_bytes, key_id = self._parse_public_key(public_key_base64)
        self._key_id = key_id
        self._verify_key = VerifyKey(public_key_bytes)

    @staticmethod
    def _parse_public_key(public_key_base64: str) -> tuple[bytes, bytes]:
        """
        Parse a minisign public key from base64 format.

        Format: 2 bytes algorithm ID + 8 bytes key ID + 32 bytes public key

        Returns:
            Tuple of (public_key_bytes, key_id).

        Raises:
            ValueError: If the key format is invalid.
        """
        try:
            decoded = base64.b64decode(public_key_base64.strip())
        except Exception as e:
            raise ValueError(f"Invalid base64 encoding in public key: {e}") from e

        expected_length = _ALGORITHM_ID_BYTES + _KEY_ID_BYTES + _ED25519_PUBLIC_KEY_BYTES
        if len(decoded) != expected_length:
            raise ValueError(
                f"Invalid public key length: expected {expected_length}, got {len(decoded)}"
            )

        # Verify algorithm ID
        algorithm_id = decoded[:_ALGORITHM_ID_BYTES]
        if algorithm_id != _ED25519_PUBKEY_ALGORITHM_ID:
            raise ValueError(
                f"Invalid algorithm ID: expected Ed25519, got {algorithm_id.hex()}"
            )

        # Extract the 8-byte key ID (right after algorithm ID)
        key_id = decoded[_ALGORITHM_ID_BYTES : _ALGORITHM_ID_BYTES + _KEY_ID_BYTES]

        # Extract the 32-byte Ed25519 public key (after key ID)
        public_key = decoded[_ALGORITHM_ID_BYTES + _KEY_ID_BYTES :]

        return public_key, key_id

    @staticmethod
    def _parse_signature(signature_string: str) -> tuple[bytes, bytes, bool]:
        """
        Parse a minisign signature from the multi-line format.

        Format:
            untrusted comment: <comment>
            <base64-encoded signature>
            trusted comment: <comment>
            <base64-encoded global signature>

        We only use the second line (the signature itself).

        Returns:
            Tuple of (key_id, signature, is_prehashed).

        Raises:
            ValueError: If signature parsing fails.
        """
        lines = signature_string.strip().split("\n")
        if len(lines) < 2:
            raise ValueError("Invalid signature format: expected at least 2 lines")

        # The signature is on the second line
        signature_base64 = lines[1].strip()

        try:
            decoded = base64.b64decode(signature_base64)
        except Exception as e:
            raise ValueError(f"Invalid base64 encoding in signature: {e}") from e

        expected_length = _ALGORITHM_ID_BYTES + _KEY_ID_BYTES + _ED25519_SIGNATURE_BYTES
        if len(decoded) != expected_length:
            raise ValueError(
                f"Invalid signature length: expected {expected_length}, got {len(decoded)}"
            )

        # Check algorithm ID - "ED" = prehashed, "Ed" = legacy non-prehashed
        algorithm_id = decoded[:_ALGORITHM_ID_BYTES]
        if algorithm_id == _ED25519_SIG_ALGORITHM_ID:
            is_prehashed = True  # "ED" - prehashed with BLAKE2b
        elif algorithm_id == _ED25519_PUBKEY_ALGORITHM_ID:
            is_prehashed = False  # "Ed" - legacy non-prehashed
        else:
            raise ValueError(
                f"Invalid algorithm ID in signature: expected Ed25519, got {algorithm_id.hex()}"
            )

        # Extract key ID
        key_id = decoded[_ALGORITHM_ID_BYTES : _ALGORITHM_ID_BYTES + _KEY_ID_BYTES]

        # Extract signature
        signature = decoded[_ALGORITHM_ID_BYTES + _KEY_ID_BYTES :]

        return key_id, signature, is_prehashed

    def verify(self, data: bytes, signature_string: str) -> bool:
        """
        Verify a minisign signature against data.

        Args:
            data: The data that was signed.
            signature_string: The minisign signature (multi-line format).

        Returns:
            True if the signature is valid, False otherwise.

        Raises:
            ValueError: If signature parsing fails.
        """
        sig_key_id, signature, is_prehashed = self._parse_signature(signature_string)

        # Verify key ID matches
        if sig_key_id != self._key_id:
            return False

        # Minisign "ED" signatures are prehashed - compute BLAKE2b-512 hash first
        # This matches SIGALG_PREHASHED in the minisign crate
        if is_prehashed:
            data_to_verify = hashlib.blake2b(data, digest_size=64).digest()
        else:
            data_to_verify = data

        # Verify the signature using Ed25519
        try:
            self._verify_key.verify(data_to_verify, signature)
            return True
        except BadSignatureError:
            return False
